### ISSUE TITLE
Deprecate emptyLineBefore option in declaration-block-properties-order

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#1308]
Feature to be moved into a plugin (e.g. `block-structure`). Include request for maintainer in deprecation notice.

Anyone fancy giving a home this plugin?

**Comment by evilebottnawi** [comment]
@jeddy3 Many people like to sort properties and padding between them as they want, I do not think it's a good idea.

**Comment by jeddy3** [comment]
Hopefully that means that someone will step-up and offer to maintain the plugin outside of the stylelint org. There are [a](https://github.com/stylelint/stylelint/issues/973) [couple](https://github.com/stylelint/stylelint/issues/705) [of](https://github.com/stylelint/stylelint/issues/752) of enhancements that could be housed in such a plugin.

**Comment by evilebottnawi** [comment]
@jeddy3 I think this rule is to leave this for what it is in the core. Unfortunately, there is no one there to maintaining it is not in the core :cry:. For `at-rule` and `pseudo` order i think we can have other rules (example `declaration-order` or better name). Almost all the other css linters have same options and after remove this options, we again received a request to add this into the core again.

**Comment by jeddy3** [comment]
> Unfortunately, there is no one there to maintaining it is not in the core

If people need it they will maintain it. The issue is that it's an open-ended rule. It'll never be "finished". If it's a plugin then people can more easily fork and add features for their needs etc. 

> Almost all the other css linters have same options and after remove this options, we again received a request to add this into the core.

I think that's a legacy of, up until recently, non of the other linters supporting plugins. We have this feature. It's powerful feature. One we should use to make this project sustainable.

**Comment by evilebottnawi** [comment]
@jeddy3 Why will never be "finished"?

**Comment by davidtheclark** [comment]
As soon as this rule stepped away from the basic use-case (which it did with groups), I think it opened the door to unlimited preferences. Some people will want their unspecified properties on top, others on bottom; some alphabetized, some not; some with spaces between groups; others without; some with spaces between groups and unspecified properties; others without; some with vendor-prefixes alphabetically ordered; others some other specific order; others with no specific order enforced; etc. etc. That's what we mean by the fact that it can't be "finished": it's just too subjective whether the rule is doing its job "well" or not.

That said, we do have this rule in core and haven't had complaints about it for a while. So I'd be okay with leaving it in core and just keeping it in mind as a warning. I think we *should* have relegated it to a plugin from the beginning, but since we didn't maybe it's too much to take what we have away. 

I don't know, it's a tough call.

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark We reserve rule as is, or deprecate "group objects"?

**Comment by evilebottnawi** [comment]
@davidtheclark i think we can finish a rule, in order to fully cover all the required options are not many, perhaps the rule will not be easy, but who said it would be easy

**Comment by jeddy3** [comment]
> I don't know, it's a tough call.

Agreed.

>  I think we should have relegated it to a plugin from the beginning, but since we didn't maybe it's too much to take what we have away.

I think the rule was built before we even had a plugin system :)

> We reserve rule as is, or deprecate "group objects"?

I think we should deprecate and nip this one in the bud before it festers. I think it'll help keep the project sustainable, and I strongly believe that's our most pressing concern here.

There are three outstanding issues related to this rule: https://github.com/stylelint/stylelint/issues/973, https://github.com/stylelint/stylelint/issues/705, and https://github.com/stylelint/stylelint/issues/752.
Two of which are related to the rule handling groups and one of which is to do with empty line between the groups, which isn't at all related to the order of properties.

**Comment by evilebottnawi** [comment]
@jeddy3 agree, lets deprecate "group objects".

**Comment by dan-gamble** [comment]
I personally would hate to see this removed, it's my biggest and main pull to Stylelint but i know the rule has caused you a lot of problems and like David said it has got very out of hand because it's stepped away from it's basic use case.

If this was moved into an exterior project / plugin i would probably start looking into `postcss-sorting` to fufill this need.

**Comment by jeddy3** [comment]
> I personally would hate to see this removed, it's my biggest and main pull to Stylelint

You will still able to use this feature within the stylelint eco-system. We've not suggesting disallowing the feature, only moving it of the stylelint org. Would a `statement-block` plugin be something you'd like to champion and give a home to under your github user?

> If this was moved into an exterior project / plugin i would probably start looking into postcss-sorting to fufill this need.

That's one of the advantages of moving these advanced features to a plugin... you can adapt the rule configuration object to match that of `postcss-sorting`, so that you can lint *and* autofix using the same config! :)

**Comment by dan-gamble** [comment]
Honestly as much as i'd like to i just honestly don't have the time to maintain something like this, my `postcss-font-awesome` plugin hasn't seen anything since it's conception and that's tiny.

It'd be awesome to see it as a plugin i'm just sure how vibrant the plugin ecosystem for Stylelint is yet

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark propose to postpone the removal and 7 versions it clear that the rule should be done and how to implement it better, in fact, we can get a lot of angry feedback because of this and lose part of the community.

**Comment by evilebottnawi** [comment]
I think this rule can be `finished`.

**Comment by dan-gamble** [comment]
Do you guys have any kind of data to know how often groups are used vs not?

**Comment by evilebottnawi** [comment]
Most people I know and who use `stylelint` use groups BTW.

**Comment by jeddy3** [comment]
> It'd be awesome to see it as a plugin i'm just sure how vibrant the plugin ecosystem for Stylelint is yet

We've not really given it a chance yet. I strongly think we do need to lean on it to make this project sustainable.

> I think this rule can be finished + Do you guys have any kind of data to know how often groups are used vs not?

It's more about *unambiguous rule names in core*. `declaration-block-properties-order` should only check for a specific *order of properties in each declaration block*. All the other currently-implemented rules names are unambiguous. 

`block-structure` would make a great plugin. "structure" is ambiguous enough to a incorporate order *of all things*, groups, whitespace etc. You can align it with `postcss-sorting` or one of the other autofixing structure PostCSS plugins out there etc.

> Most people I know and who use stylelint use groups BTW.

@evilebottnawi Do you groups? If so, would you like house and champion this plugin? We could get everything in place for version 7 and so the migration shouldn't be painful.

**Comment by dan-gamble** [comment]
> We've not really given it a chance yet. I strongly think we do need to lean on it to make this project sustainable.

Fully agree. With the rate rules are being added it does seem like it would make sense to trim some of the fat and move them into plugins, i just feel at the moment a lot of things that get moved into plugins would become almost redundant and unused but there is a time where you need to start pushing the plugin side of Stylelint as it's a very strong benefit of Stylelint.

> `block-structure` would make a great plugin

I agree again would be a nice base to build on and hopefully remove a lot of the complexity of the rule if it was split into smaller groups.

**Comment by evilebottnawi** [comment]
@jeddy3 no in my commands we not use groups :smile:, But for others it can be painful, I just leave it until the offer as he is and apart to open the issue for debate about the best organization or its possible simplification

**Comment by jeddy3** [comment]
>  i just feel at the moment a lot of things that get moved into plugins would become almost redundant and unused 

If grouping things is popular then it's the perfect catalyst for the plugin eco-system.

> but there is a time where you need to start pushing the plugin side of Stylelint as it's a very strong benefit of Stylelint.

I believe *now* is that time. Not just for the user's sake, but for the sustainability of the project after next month.

**Comment by dan-gamble** [comment]
> I believe now is that time. Not just for the user's sake, but for the sustainability of the project after next month. 👍

**Comment by davidtheclark** [comment]
Thanks for all the discussion above.

I think we should deprecate the grouping feature of the rule. It makes sense to me because of reasons stated above — it is unwieldy, without clear boundaries, and the source of a disproportionate number of issues.

It's also an excellent candidate for an independent project, whether as a stylelint plugin or a postcss plugin like `postcss-sorting`. There's really an indefinite number of features that people could build into a "sorting" tool/rule (which is why it can't be "finished" I think).

I understand that people like it and will be disappointed. Those people that like it should create a plugin and maintain it. If nobody who likes it is willing to step up and create the plugin, that's on them: you can't always get everything you want for free.

**Comment by jeddy3** [comment]
Perfect. Then we're in agreement.

I'll try to get the deprecation warnings in before we pack up tools, to give people as much time as possible to build alternatives before we start work on `v7`.

**Comment by jeddy3** [comment]
>  I think we should deprecate the grouping feature of the rule. It makes sense to me because of reasons stated above — it is unwieldy, without clear boundaries, and the source of a disproportionate number of issues.

While working on deprecating the "group objects" configuration feature this morning I've come to realise that I wrongly assumed the "group objects" feature was *only* to do with whitespace, whereas I now realise it is also used to legitimately define a *semi-flexible property-order* (using the `order: "flexible"` option).

This discovery is a *good thing*. I think it means that both the transition to `7.0.0` is going to be smoother and we still get to define clear boundaries (which means we still get to push the outstanding issues to a plugin) e.g.

- https://github.com/stylelint/stylelint/issues/973 - concerned with the order of things *that aren't properties*.
- https://github.com/stylelint/stylelint/issues/705 - concerned with *whitespace between declarations*.
- https://github.com/stylelint/stylelint/issues/752 - concerned with *whitespace between declarations*.

In summary, `declaration-block-properties-order` is only concerned with *the order of properties within a declaration-block*. It is **not** concerned with *whitespace* nor *the order of non-properties*.

So, I believe we should only deprecate the `emptyLineBefore` option. The other options (flexible groups and `unspecified`) are legitimate.

I've [a pending PR](https://github.com/stylelint/stylelint/issues/1350) that updates the release planning doc to reflect this. Have a look through that and see what you think. Hopefully, we've stumbled onto the sweet spot that defines a clear purpose and scope for this rule, while offering a smoother transition to `7.0.0`.

**Comment by davidtheclark** [comment]
Ok!

**Comment by jeddy3** [comment]
Done in https://github.com/stylelint/stylelint/pull/1353
